website/blog/2018-06-25-on-consuming-and-publishing-es6+-packages.md

```diff
@@ -0,0 +1,218 @@
+
+So actually publishing it in a library may cause even more issues. Because a TC39 proposal (even at Stage 3) has a possibility of changing, it means you will inevitability have to change the code.
+
+At least if you ship the compiled version, it will still work and the library maintainer can change the output so that it works the same as before. Shipping the uncompiled version means that anyone consuming a package will necessarily have to have a build step to use it, and will have to have the same version of Babel as you. This is no different than using TS/JSX/Flow which you wouldn't publish either.
+
+TODO: Or a proposal might just stall because people are busy, etc. Decorators are a great example of this.
+
+### Conflating ESM and ES6+
+
+When you do `import foo from "foo"` or `require("foo")` and `foo` doesn't have an `index.js`, it will resolve to the `main` field in the `package.json` of the module.
+
+Some tools like Rollup/Webpack will also read from another field called `module` (previously `jsnext:main`). It uses this to instead resolve to the ESM file.
+
+- An example with [`redux`](https://github.com/reactjs/redux)
+
+```js
+// redux package.json
+{
+  ...
+  "main": "lib/redux.js", // ES5 + CJS
+  "module": "es/redux.js", // ES5 + ESM
+}
+```
+
+This was introduced so that users could consume ES2015 modules (ESM).
+
+However, the sole intention of this field is ESM, not anything else. The [Rollup docs](https://github.com/rollup/rollup/wiki/pkg.module#wait-it-just-means-import-and-export--not-other-future-javascript-features) make it clear that it's not intended for future JavaScript syntax.
+
+What this means is that we may need another way to handle ES6+ even though some libraries may choose to do this anyway.
+
+A common suggestion is for libraries to start publishing ES2015 under another field like `es2015`, e.g. `"es2015": "es2015/redux.js"`.
+
+```js
+// @angular/core package.json
+{
+  "main": "./bundles/core.umd.js",
+  "module": "./fesm5/core.js",
+  "es2015": "./fesm2015/core.js",
+  "esm5": "./esm5/core.js",
+  "esm2015": "./esm2015/core.js",
+  "fesm5": "./fesm5/core.js",
+  "fesm2015": "./fesm2015/core.js",
+}
+```
+
+This works for ES2015 but the question next is what about ES2016? Are we supposed to create a new folder for each year and a new field in `package.json`? That seems unsustainable, and will continue to produce larger `node_modules`.
+
+I feel like this is similar to how we had a Babel preset for each year and ended up creating `preset-latest` which included all the yearly plugins (we removed that soon after in favor of `preset-env`).
+
+Having a field like `es-latest` or `esnext` may not be helpful either because we want the library versions to be fixed yet changing. And providing only the source may not always be helpful if you use non-standard syntax, TS, Flow, JSX, etc which might have different behavior based on compiler version. Maybe we need some changes on the `npm` side?
```

Comment:
You could clarify what you mean by "fixed yet changing" a bit more.

> Having a field named `es-latest` or `esnext` would not help, either. What, for instance, is the "latest" version of JavaScript? The answer changes depending on the point in time you authored the code!
>
> What about providing just the source? That would have similar problems if you used non-standard syntax, TS, Flow, JSX, etc., since they may yield different output depending on the versions of the compilers you used at the time you authored them.